core/services/workflows/engine.go

```diff
@@ -505,86 +537,71 @@ func (e *Engine) startExecution(ctx context.Context, executionID string, event *
-				step, ok := state.Steps[s.Ref]
-				// The step is missing from the state,
-				// which means it hasn't been processed yet.
-				// Let's mark `workflowCompleted` = false, and
-				// continue.
-				if !ok {
-					workflowCompleted = false
-					return nil
-				}
 
-				switch step.Status {
-				case store.StatusCompleted, store.StatusErrored, store.StatusCompletedEarlyExit:
-				default:
-					workflowCompleted = false
-				}
-				return nil
-			})
-			if err != nil {
-				return err
-			}
-
-			if workflowCompleted {
-				return e.finishExecution(ctx, state.ExecutionID, store.StatusCompleted)
-			}
-		}
+	workflowIsFullyProcessed, status, err := e.isWorkflowFullyProcessed(state)
+	if err != nil {
+		return err
+	}
 
-		// We haven't completed the workflow, but should we continue?
-		// If we've been executing for too long, let's time the workflow out and stop here.
-		if state.CreatedAt != nil && e.clock.Since(*state.CreatedAt) > e.maxExecutionDuration {
+	if workflowIsFullyProcessed {
+		switch status {
+		case store.StatusTimeout:
 			l.Info("execution timed out")
-			return e.finishExecution(ctx, state.ExecutionID, store.StatusTimeout)
+		case store.StatusCompleted:
+			l.Info("workflow finished")
+		case store.StatusErrored:
+			l.Info("execution errored")
+		case store.StatusCompletedEarlyExit:
+			l.Info("execution terminated early")
+			// NOTE: even though this marks the workflow as completed, any branches of the DAG
+			// that don't depend on the step that signaled for an early exit will still complete.
+			// This is to ensure that any side effects are executed consistently, since otherwise
+			// the async nature of the workflow engine would provide no guarantees.
 		}
+		return e.finishExecution(ctx, state.ExecutionID, status)
+	}
 
-		// Finally, since the workflow hasn't timed out or completed, let's
-		// check for any dependents that are ready to process.
-		for _, sd := range stepDependents {
-			e.queueIfReady(state, sd)
-		}
-	case store.StatusCompletedEarlyExit:
-		l.Info("execution terminated early")
-		// NOTE: even though this marks the workflow as completed, any branches of the DAG
-		// that don't depend on the step that signaled for an early exit will still complete.
-		// This is to ensure that any side effects are executed consistently, since otherwise
-		// the async nature of the workflow engine would provide no guarantees.
-		err := e.finishExecution(ctx, state.ExecutionID, store.StatusCompletedEarlyExit)
-		if err != nil {
-			return err
-		}
-	case store.StatusErrored:
-		l.Info("execution errored")
-		err := e.finishExecution(ctx, state.ExecutionID, store.StatusErrored)
-		if err != nil {
-			return err
-		}
+	// Finally, since the workflow hasn't timed out or completed, let's
+	// check for any dependents that are ready to process.
+	stepDependents, err := e.workflow.dependents(stepUpdate.Ref)
+	if err != nil {
+		return err
+	}
+	for _, sd := range stepDependents {
+		e.queueIfReady(state, sd)
```

Comment:
is it cleaner wrt to separation of concerns if `handleStepUpdate` isn't responsible for queue'ing dependents?

Reply:
The call to queue is really part of handling the update here.

Reply:
Aren't they separate concerns? If there was a parent orchestrator that called handleStepUpdate and then processed the dependents, that wouldn't be the case.